File: java/com/google/copybara/git/GithubPrDestination.java
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.copybara.git;

import static com.google.copybara.git.LazyGitRepository.memoized;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Preconditions;
import com.google.common.base.Strings;
import com.google.common.collect.ImmutableSetMultimap;
import com.google.common.collect.ImmutableSetMultimap.Builder;
import com.google.copybara.ChangeMessage;
import com.google.copybara.Destination;
import com.google.copybara.GeneralOptions;
import com.google.copybara.RepoException;
import com.google.copybara.TransformResult;
import com.google.copybara.ValidationException;
import com.google.copybara.git.GitDestination.CommitGenerator;
import com.google.copybara.git.GitDestination.ProcessPushOutput;
import com.google.copybara.git.GitDestination.WriterImpl;
import com.google.copybara.git.GitDestination.WriterState;
import com.google.copybara.git.github_api.CreatePullRequest;
import com.google.copybara.git.github_api.GithubApi;
import com.google.copybara.git.github_api.PullRequest;
import com.google.copybara.util.Glob;
import com.google.copybara.util.console.Console;
import java.io.IOException;
import java.util.UUID;
import javax.annotation.Nullable;

/**
 * A destination for creating/updating Github Pull Requests.
 */
public class GithubPrDestination implements Destination<GitRevision> {

  private final String url;
  private final String destinationRef;
  private final GeneralOptions generalOptions;
  private final GithubOptions githubOptions;
  private final GitDestinationOptions destinationOptions;
  private final GithubDestinationOptions githubDestinationOptions;
  private final GitOptions gitOptions;
  private final CommitGenerator commitGenerator;
  private final ProcessPushOutput processPushOutput;
  private final Iterable<GitIntegrateChanges> integrates;
  private final boolean effectiveSkipPush;
  private final LazyGitRepository localRepo;

  public GithubPrDestination(String url, String destinationRef, GeneralOptions generalOptions,
      GithubOptions githubOptions,
      GitDestinationOptions destinationOptions, GithubDestinationOptions githubDestinationOptions,
      GitOptions gitOptions,
      boolean skipPush, CommitGenerator commitGenerator, ProcessPushOutput processPushOutput,
      Iterable<GitIntegrateChanges> integrates) {
    this.url = Preconditions.checkNotNull(url);
    this.destinationRef = Preconditions.checkNotNull(destinationRef);
    this.generalOptions = Preconditions.checkNotNull(generalOptions);
    this.githubOptions = Preconditions.checkNotNull(githubOptions);
    this.destinationOptions = Preconditions.checkNotNull(destinationOptions);
    this.githubDestinationOptions = Preconditions.checkNotNull(githubDestinationOptions);
    this.gitOptions = Preconditions.checkNotNull(gitOptions);
    this.commitGenerator = Preconditions.checkNotNull(commitGenerator);
    this.processPushOutput = Preconditions.checkNotNull(processPushOutput);
    this.integrates = Preconditions.checkNotNull(integrates);
    this.effectiveSkipPush = skipPush || destinationOptions.skipPush;
    this.localRepo = memoized(ignored -> destinationOptions.localGitRepo(url));
  }

  @Override
  public String getType() {
    return "git.github_pr_destination";
  }

  @Override
  public ImmutableSetMultimap<String, String> describe(Glob originFiles) {
    ImmutableSetMultimap.Builder<String, String> builder =
        new ImmutableSetMultimap.Builder<String, String>()
            .put("type", getType())
            .put("name", url)
            .put("destination_ref", destinationRef);
    if (effectiveSkipPush) {
      builder.put("skip_push", "True");
    }
    return builder.build();
  }

  @Override
  public Writer<GitRevision> newWriter(Glob destinationFiles, boolean dryRun,
      @Nullable String groupId, @Nullable Writer<GitRevision> oldWriter)
      throws ValidationException {
    WriterImpl gitOldWriter = (WriterImpl) oldWriter;

    boolean effectiveSkipPush = GithubPrDestination.this.effectiveSkipPush || dryRun;

    GithubWriterState state;
    String pushBranchName = branchFromGroupId(groupId);
    if (oldWriter != null && gitOldWriter.skipPush == effectiveSkipPush) {
      state = (GithubWriterState) ((WriterImpl) oldWriter).state;
    } else {
      state = new GithubWriterState(localRepo,
          destinationOptions.localRepoPath != null
              ? pushBranchName
              : "copybara/push-" + UUID.randomUUID() + (dryRun ? "-dryrun" : ""));
    }

    return new WriterImpl<GithubWriterState>(destinationFiles, effectiveSkipPush, url,
        destinationRef, pushBranchName,
        generalOptions, commitGenerator, processPushOutput,
        state, /*nonFastForwardPush=*/true, integrates,
        destinationOptions.lastRevFirstParent,
        destinationOptions.ignoreIntegrationErrors,
        destinationOptions.localRepoPath,
        destinationOptions.committerName,
        destinationOptions.committerEmail,
        destinationOptions.rebaseWhenBaseline(),
        gitOptions.visitChangePageSize) {
      @Override
      public WriterResult write(TransformResult transformResult, Console console)
          throws ValidationException, RepoException, IOException {
        WriterResult result = super.write(transformResult, console);

        if (effectiveSkipPush || state.pullRequestNumber != null) {
          return result;
        }

        if (!githubDestinationOptions.createPullRequest) {
          console.infoFmt("Please create a PR manually following this link: %s/compare/%s...%s"
                  + " (Only needed once)",
              asHttpsUrl(), destinationRef, pushBranchName);
          state.pullRequestNumber = -1L;
          return result;
        }

        GithubApi api = githubOptions.getApi(GithubUtil.getProjectNameFromUrl(url));
        for (PullRequest pr : api.getPullRequests(getProjectName())) {
          if (pr.isOpen() && pr.getHead().getRef().equals(pushBranchName)) {
            console.infoFmt("Pull request for branch %s already exists as %s/pull/%s",
                pushBranchName, asHttpsUrl(), pr.getNumber());
            if (!pr.getBase().getRef().equals(destinationRef)) {
              // TODO(malcon): Update PR or create a new one?
              console.warnFmt("Current base branch '%s' is different from the PR base branch '%s'",
                  destinationRef, pr.getBase().getRef());
            }
            return result;
          }
        }
        ChangeMessage msg = ChangeMessage.parseMessage(transformResult.getSummary());
        PullRequest pr = api.createPullRequest(getProjectName(),
            new CreatePullRequest(msg.firstLine(), msg.getText(),
                pushBranchName, destinationRef));
        console.infoFmt("Pull Request %s/pull/%s created using branch '%s'.", asHttpsUrl(),
            pr.getNumber(), pushBranchName);
        state.pullRequestNumber = pr.getNumber();
        return result;
      }
    };
  }


  private String asHttpsUrl() throws ValidationException {
    return "https://github.com/" + getProjectName();
  }

  @VisibleForTesting
  String getProjectName() throws ValidationException {
    return GithubUtil.getProjectNameFromUrl(url);
  }

  private String branchFromGroupId(@Nullable String groupId) throws ValidationException {
    if (!Strings.isNullOrEmpty(githubDestinationOptions.destinationPrBranch)) {
      return githubDestinationOptions.destinationPrBranch;
    }
    // We could do more magic here with the change identity. But this is already complex so we
    // require  a group identity either provided by the origin or the workflow (Will be implemented
    // later.
    ValidationException.checkCondition(groupId != null,
        "git.github_pr_destination is incompatible with the current origin. Origin has to be"
            + " able to provide the group identity or use '%s' flag",
        GithubDestinationOptions.GITHUB_DESTINATION_PR_BRANCH);
    return groupId.replaceAll("[^A-Za-z0-9_-]", "_");
  }

  @Override
  public String getLabelNameWhenOrigin() {
    return GitRepository.GIT_ORIGIN_REV_ID;
  }

  private static class GithubWriterState extends WriterState {

    @Nullable
    Long pullRequestNumber;

    GithubWriterState(LazyGitRepository localRepo, String localBranch) {
      super(localRepo, localBranch);
    }
  }
}

File: src/Copybara.Core/Git/GitHubPrDestination.cs
/*
 * Copyright (C) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System.Collections.Immutable;
using Copybara.Checks;
using Copybara.Common;
using Copybara.Config;
using Copybara.Effect;
using Copybara.Exceptions;
using Copybara.Git.GitHub.Api;
using Copybara.Git.GitHub.Util;
using Copybara.Revision;
using Copybara.TemplateToken;
using Copybara.Util;
using Console = Copybara.Util.Console.Console;
using GitHubApiClient = Copybara.Git.GitHub.Api.GitHubApi;

namespace Copybara.Git;

/// <summary>
/// A destination for creating/updating GitHub Pull Requests. Port of
/// <c>com.google.copybara.git.GitHubPrDestination</c>.
/// </summary>
public class GitHubPrDestination : IDestination<GitRevision>
{
    // Mirrors GitModule.PRIMARY_BRANCHES.
    private static readonly ImmutableHashSet<string> PrimaryBranches =
        ImmutableHashSet.Create("master", "main");

    private readonly string _url;
    private readonly string _destinationRef;
    private readonly string? _prBranch;
    private readonly bool _partialFetch;
    private readonly bool _draft;
    private readonly bool _primaryBranchMigrationMode;

    private readonly GeneralOptions _generalOptions;
    private readonly GitHubOptions _gitHubOptions;
    private readonly GitDestinationOptions _destinationOptions;
    private readonly GitHubDestinationOptions _gitHubDestinationOptions;
    private readonly GitOptions _gitOptions;
    private readonly GitHubPrWriteHook _writeHook;
    private readonly IEnumerable<GitIntegrateChanges> _integrates;
    private readonly string? _title;
    private readonly string? _body;
    private readonly bool _updateDescription;
    private readonly GitHubHost _ghHost;
    private readonly IChecker? _checker;
    private readonly LazyResourceLoader<GitRepository> _localRepo;
    private readonly ConfigFile _mainConfigFile;
    private readonly IReadOnlyList<string> _assignees;
    private readonly IChecker? _endpointChecker;

    private string? _resolvedDestinationRef;
    internal CredentialFileHandler? Credentials;

    public GitHubPrDestination(
        string url,
        string destinationRef,
        string? prBranch,
        bool partialFetch,
        bool draft,
        GeneralOptions generalOptions,
        GitHubOptions gitHubOptions,
        GitDestinationOptions destinationOptions,
        GitHubDestinationOptions gitHubDestinationOptions,
        GitOptions gitOptions,
        GitHubPrWriteHook writeHook,
        IEnumerable<GitIntegrateChanges> integrates,
        string? title,
        string? body,
        IReadOnlyList<string> assignees,
        ConfigFile mainConfigFile,
        IChecker? endpointChecker,
        bool updateDescription,
        GitHubHost ghHost,
        bool primaryBranchMigrationMode,
        IChecker? checker,
        CredentialFileHandler? credentials)
    {
        _url = Preconditions.CheckNotNull(url);
        _destinationRef = Preconditions.CheckNotNull(destinationRef);
        _prBranch = prBranch;
        _partialFetch = partialFetch;
        _draft = draft;
        _generalOptions = Preconditions.CheckNotNull(generalOptions);
        _gitHubOptions = Preconditions.CheckNotNull(gitHubOptions);
        _destinationOptions = Preconditions.CheckNotNull(destinationOptions);
        _gitHubDestinationOptions = Preconditions.CheckNotNull(gitHubDestinationOptions);
        _gitOptions = Preconditions.CheckNotNull(gitOptions);
        _writeHook = Preconditions.CheckNotNull(writeHook);
        _integrates = Preconditions.CheckNotNull(integrates);
        _title = title;
        _assignees = assignees;
        _body = body;
        _updateDescription = updateDescription;
        _ghHost = Preconditions.CheckNotNull(ghHost);
        _checker = checker;
        Credentials = credentials;
        _localRepo = LazyResourceLoader.Memoized<GitRepository>(
            _ => destinationOptions.LocalGitRepo(url, credentials));
        _mainConfigFile = Preconditions.CheckNotNull(mainConfigFile);
        _endpointChecker = endpointChecker;
        _primaryBranchMigrationMode = primaryBranchMigrationMode;
    }

    public string GetType() => "git.github_pr_destination";

    public ImmutableListMultimap<string, string> Describe(Glob? destinationFiles)
    {
        var builder = ImmutableListMultimap<string, string>.CreateBuilder();
        builder.Put("type", GetType());
        builder.Put("url", _url);
        builder.Put("destination_ref", _destinationRef);
        builder.Put("primaryBranchMigrationMode", _primaryBranchMigrationMode.ToString());
        if (_checker != null)
        {
            builder.Put("checker", _checker.GetType().FullName ?? _checker.GetType().Name);
        }
        if (destinationFiles != null
            && !destinationFiles.Roots().IsEmpty
            && !destinationFiles.Roots().Contains(""))
        {
            builder.PutAll("root", destinationFiles.Roots());
        }
        return builder.Build();
    }

    public IReadOnlyList<ImmutableListMultimap<string, string>> DescribeCredentials() =>
        Credentials == null
            ? ImmutableArray<ImmutableListMultimap<string, string>>.Empty
            : GitDescribeCredentials.Convert(Credentials.DescribeCredentials());

    public IDestination<GitRevision>.IWriter<GitRevision> NewWriter(WriterContext writerContext)
    {
        string prBranch =
            GetPullRequestBranchName(
                writerContext.GetOriginalRevision(),
                writerContext.GetWorkflowName(),
                writerContext.GetWorkflowIdentityUser());
        GitHubPrWriteHook gitHubPrWriteHook = _writeHook.WithUpdatedPrBranch(prBranch);

        var state = new GitHubWriterState(
            _localRepo,
            _destinationOptions.LocalRepoPath != null
                ? prBranch
                : "copybara/push-"
                    + Guid.NewGuid()
                    + (writerContext.IsDryRun() ? "-dryrun" : ""));

        return new GitHubPrWriterImpl(
            this,
            writerContext,
            prBranch,
            gitHubPrWriteHook,
            state);
    }

    internal string AsHttpsUrl() => _ghHost.ProjectAsUrl(GetProjectName());

    internal string GetProjectName() => _ghHost.GetProjectNameFromUrl(_url);

    public bool IsUpdateDescription() => _updateDescription;

    public IEnumerable<GitIntegrateChanges> GetIntegrates() => _integrates;

    private string GetPullRequestBranchName(
        IRevision? changeRevision, string workflowName, string workflowIdentityUser)
    {
        if (!string.IsNullOrEmpty(_gitHubDestinationOptions.DestinationPrBranch))
        {
            return _gitHubDestinationOptions.DestinationPrBranch!;
        }
        string? contextReference = changeRevision?.ContextReference();
        ValidationException.CheckCondition(
            contextReference != null,
            "git.github_pr_destination is incompatible with the current origin. Origin has to be"
                + " able to provide the contextReference or use '{0}' flag",
            GitHubDestinationOptions.GitHubDestinationPrBranch);
        string? branchNameFromUser = GetCustomBranchName(contextReference!);
        string branchName =
            branchNameFromUser
            ?? Identity.ComputeIdentity(
                "OriginGroupIdentity",
                contextReference!,
                workflowName,
                _mainConfigFile.GetIdentifier(),
                workflowIdentityUser);
        return GitHubUtil.GetValidBranchName(branchName);
    }

    private string? GetCustomBranchName(string contextReference)
    {
        if (_prBranch == null)
        {
            return null;
        }
        try
        {
            return new LabelTemplate(_prBranch)
                .Resolve(e => e == "CONTEXT_REFERENCE" ? contextReference : _prBranch);
        }
        catch (LabelTemplate.LabelNotFoundException e)
        {
            throw new ValidationException(
                "Cannot find some labels in the GitHub PR branch name field: " + e.Message, e);
        }
    }

    public string GetLabelNameWhenOrigin() => GitRepository.GitOriginRevId;

    internal string GetDestinationRef()
    {
        if (!_primaryBranchMigrationMode || !PrimaryBranches.Contains(_destinationRef))
        {
            return _destinationRef;
        }
        if (_resolvedDestinationRef == null)
        {
            try
            {
                GitRepository repo = _localRepo.Load(_generalOptions.GetConsole());
                string? primaryBranch = repo.GetPrimaryBranch(_url);
                _resolvedDestinationRef = primaryBranch ?? _destinationRef;
            }
            catch (RepoException e)
            {
                _generalOptions.GetConsole().WarnFmt("Error detecting primary branch: {0}", e.Message);
                _resolvedDestinationRef = _destinationRef;
            }
        }
        return _resolvedDestinationRef;
    }

    private sealed class GitHubWriterState : GitDestination.WriterState
    {
        internal long? PullRequestNumber;

        internal GitHubWriterState(
            LazyResourceLoader<GitRepository> localRepo, string localBranch)
            : base(localRepo, localBranch)
        {
        }
    }

    private sealed class GitHubPrWriterImpl : GitDestination.WriterImpl<GitHubWriterState>
    {
        private readonly GitHubPrDestination _destination;
        private readonly WriterContext _writerContext;
        private readonly string _prBranch;
        private readonly GitHubWriterState _state;

        internal GitHubPrWriterImpl(
            GitHubPrDestination destination,
            WriterContext writerContext,
            string prBranch,
            GitHubPrWriteHook gitHubPrWriteHook,
            GitHubWriterState state)
            : base(
                writerContext.IsDryRun(),
                destination._url,
                destination.GetDestinationRef(),
                prBranch,
                destination._partialFetch,
                tagNameTemplate: null,
                tagMsgTemplate: null,
                destination._generalOptions,
                destination._gitOptions,
                gitHubPrWriteHook,
                state,
                nonFastForwardPush: true,
                destination._integrates,
                destination._destinationOptions.LastRevFirstParent,
                destination._destinationOptions.IgnoreIntegrationErrors,
                destination._destinationOptions.LocalRepoPath,
                destination._destinationOptions.CommitterName,
                destination._destinationOptions.CommitterEmail,
                destination._destinationOptions.RebaseWhenBaseline(),
                destination._gitOptions.VisitChangePageSize,
                destination._gitOptions.GitTagOverwrite,
                destination._checker,
                destination._destinationOptions,
                destination.Credentials)
        {
            _destination = destination;
            _writerContext = writerContext;
            _prBranch = prBranch;
            _state = state;
        }

        public override IReadOnlyList<DestinationEffect> Write(
            TransformResult transformResult, Glob destinationFiles, Console console)
        {
            var result = new List<DestinationEffect>(
                base.Write(transformResult, destinationFiles, console));
            if (_writerContext.IsDryRun() || _state.PullRequestNumber != null)
            {
                return result;
            }

            if (!_destination._gitHubDestinationOptions.CreatePullRequest)
            {
                console.InfoFmt(
                    "Please create a PR manually following this link: {0}/compare/{1}...{2}"
                        + " (Only needed once)",
                    _destination.AsHttpsUrl(),
                    _destination.GetDestinationRef(),
                    _prBranch);
                _state.PullRequestNumber = -1L;
                return result;
            }

            GitHubApiClient api =
                _destination._gitHubOptions.NewGitHubRestApi(
                    _destination._ghHost.GetHost(),
                    _destination.GetProjectName(),
                    null,
                    _destination.Credentials,
                    console);

            var pullRequests =
                api.GetPullRequestsAsync(
                    _destination.GetProjectName(),
                    GitHubApiClient.PullRequestListParams.Default.WithHead(
                        $"{_destination._ghHost.GetUserNameFromUrl(_destination._url)}:{_prBranch}"))
                    .GetAwaiter().GetResult();

            ChangeMessage msg = ChangeMessage.ParseMessage(transformResult.GetSummary().Trim());

            string? title =
                _destination._title == null
                    ? msg.FirstLine()
                    : LabelFinder.MapLabels(
                        transformResult.GetLabelFinder(), _destination._title, "title");

            string prBody =
                _destination._body == null
                    ? msg.ToString()
                    : LabelFinder.MapLabels(
                        transformResult.GetLabelFinder(), _destination._body, "body");
            var assignees =
                LabelFinder.MapLabels(
                    transformResult.GetLabelFinder(), _destination._assignees.ToList());

            foreach (var pr in pullRequests)
            {
                if (pr.GetHead()!.GetRef() == _prBranch)
                {
                    if (!pr.IsOpen())
                    {
                        console.WarnFmt(
                            "Pull request for branch {0} already exists as {1}/pull/{2}, but is"
                                + " closed - reopening.",
                            _prBranch,
                            _destination.AsHttpsUrl(),
                            pr.GetNumber());
                        api.UpdatePullRequestAsync(
                            _destination.GetProjectName(),
                            pr.GetNumber(),
                            new UpdatePullRequest(null, null, UpdatePullRequestState.OPEN))
                            .GetAwaiter().GetResult();
                    }
                    else
                    {
                        console.InfoFmt(
                            "Pull request for branch {0} already exists as {1}/pull/{2}",
                            _prBranch, _destination.AsHttpsUrl(), pr.GetNumber());
                    }
                    if (pr.GetBase()!.GetRef() != _destination.GetDestinationRef())
                    {
                        console.WarnFmt(
                            "Current base branch '{0}' is different from the PR base branch '{1}'",
                            _destination.GetDestinationRef(), pr.GetBase()!.GetRef());
                    }
                    if (_destination._updateDescription)
                    {
                        ValidationException.CheckCondition(
                            !string.IsNullOrEmpty(title),
                            "Pull Request title cannot be empty. Either use 'title' field in"
                                + " git.github_pr_destination or modify the message to not be empty");
                        api.UpdatePullRequestAsync(
                            _destination.GetProjectName(),
                            pr.GetNumber(),
                            new UpdatePullRequest(title, prBody, state: null))
                            .GetAwaiter().GetResult();
                    }
                    result.Add(
                        new DestinationEffect(
                            DestinationEffect.EffectType.UPDATED,
                            $"Pull Request {pr.GetHtmlUrl()} updated",
                            transformResult.GetChanges().GetCurrent().Cast<OriginRef>().ToList(),
                            new DestinationEffect.DestinationRef(
                                pr.GetNumber().ToString(), "pull_request", pr.GetHtmlUrl())));
                    return result;
                }
            }

            ValidationException.CheckCondition(
                !string.IsNullOrEmpty(title),
                "Pull Request title cannot be empty. Either use 'title' field in"
                    + " git.github_pr_destination or modify the message to not be empty");

            PullRequest newPr =
                api.CreatePullRequestAsync(
                    _destination.GetProjectName(),
                    new CreatePullRequest(
                        title!, prBody, _prBranch, _destination.GetDestinationRef(),
                        _destination._draft))
                    .GetAwaiter().GetResult();
            console.InfoFmt(
                "Pull Request {0}/pull/{1} created using branch '{2}'.",
                _destination.AsHttpsUrl(), newPr.GetNumber(), _prBranch);

            if (assignees.Count != 0)
            {
                try
                {
                    api.AddAssigneesAsync(
                        _destination.GetProjectName(),
                        newPr.GetNumber(),
                        new AddAssignees(assignees.ToList()))
                        .GetAwaiter().GetResult();
                }
                catch (RepoException e)
                {
                    console.WarnFmt(
                        "Could not add all assignees ({0}) to {1}/pull/{2} with error '{3}'",
                        string.Join(", ", assignees),
                        _destination.AsHttpsUrl(),
                        newPr.GetNumber(),
                        e.Message);
                }
            }

            _state.PullRequestNumber = newPr.GetNumber();
            result.Add(
                new DestinationEffect(
                    DestinationEffect.EffectType.CREATED,
                    $"Pull Request {newPr.GetHtmlUrl()} created",
                    transformResult.GetChanges().GetCurrent().Cast<OriginRef>().ToList(),
                    new DestinationEffect.DestinationRef(
                        newPr.GetNumber().ToString(), "pull_request", newPr.GetHtmlUrl())));
            return result;
        }

        public override IEndpoint GetFeedbackEndPoint(Console console)
        {
            _destination._gitHubOptions.ValidateEndpointChecker(_destination._endpointChecker);
            return new GitHubEndPoint(
                _destination._gitHubOptions.NewGitHubApiSupplier(
                    _destination._url,
                    _destination._endpointChecker,
                    _destination.Credentials,
                    _destination._ghHost),
                _destination._url,
                console,
                _destination._ghHost,
                _destination.Credentials);
        }
    }
}
